runtime/interpreter/interpreter.go

```diff
@@ -2214,6 +2217,190 @@ func (interpreter *Interpreter) writeStored(
 	interpreter.recordStorageMutation()
 }
 
+type fromStringFunctionValue struct {
+	receiverType sema.Type
+	hostFunction *HostFunctionValue
+}
+
+// a function that attempts to create a Cadence value from a string, e.g. parsing a number from a string
+type valueParser func(*Interpreter, string) OptionalValue
+
+func newFromStringFunction(ty sema.Type, parser valueParser) fromStringFunctionValue {
+	functionType := sema.FromStringFunctionType(ty)
+
+	hostFunctionImpl := NewUnmeteredHostFunctionValue(
+		func(invocation Invocation) Value {
+			argument, ok := invocation.Arguments[0].(*StringValue)
+			if !ok {
+				// expect typechecker to catch a mismatch here
+				panic(errors.NewUnreachableError())
+			}
+			inter := invocation.Interpreter
+			return parser(inter, argument.Str)
+		},
+		functionType,
+	)
+	return fromStringFunctionValue{
+		receiverType: ty,
+		hostFunction: hostFunctionImpl,
+	}
+}
+
+// default implementation for parsing a given unsigned numeric type from a string.
+// the size provided by sizeInBytes is passed to strconv.ParseUint, ensuring that the parsed value fits in the target type.
+// input strings must not begin with a '+' or '-'.
+func unsignedIntValueParser(bitSize int, convert func(uint64) Value) valueParser {
+	expectedUsage := common.NewNumberMemoryUsage(bitSize / 8)
+
+	return func(interpreter *Interpreter, input string) OptionalValue {
+		val, err := strconv.ParseUint(input, 10, bitSize)
+		if err != nil {
+			return NewNilValue(interpreter)
+		}
+
+		converted := convert(val)
+		common.UseMemory(interpreter, expectedUsage)
+		return NewSomeValueNonCopying(interpreter, converted)
+	}
+}
+
+// default implementation for parsing a given signed numeric type from a string.
+// the size provided by sizeInBytes is passed to strconv.ParseUint, ensuring that the parsed value fits in the target type.
+// input strings may begin with a '+' or '-'.
+func signedIntValueParser(bitSize int, convert func(int64) Value) valueParser {
+	expectedUsage := common.NewCadenceBigIntMemoryUsage(bitSize / 8)
+
+	return func(interpreter *Interpreter, input string) OptionalValue {
+		val, err := strconv.ParseInt(input, 10, bitSize)
+		if err != nil {
+			return NewNilValue(interpreter)
+		}
+
+		converted := convert(val)
+		common.UseMemory(interpreter, expectedUsage)
+		return NewSomeValueNonCopying(interpreter, converted)
+	}
+}
+
+func bigIntValueParser(convert func(*big.Int) (Value, bool)) valueParser {
+	return func(interpreter *Interpreter, input string) OptionalValue {
+		// maxVal =~ 10^|input|
+		// bytes = ceil(log256(maxVal))
+		// = ceil(log256(10 ^ |input|)) via substitution
+		// = ceil(|input| * log256(10)) via power rule
+		// = ceil(|input| * 0.415241011)
+		// there's gotta be a better way of estimating memory usage lol
+
+		estimatedSize := int(math.Ceil(float64(len(input)) * 0.415241011))
+		common.UseMemory(interpreter, common.NewCadenceBigIntMemoryUsage(estimatedSize))
```

Comment:
Memory usage is deliberately an overestimate. If you're multiplying by a decimal value you can just round to the nearest whole number for convenience. In particular, I think it's fine here to just use `len(input)` as the estimated size.

Reply:
Agreed, we can just overestimate, we do not need to be accurate, as long as we do not underestimate.

One problem might be that the function does not just allocate for the result, but performs additional temporary allocations. Looking at `Int.SetString` it indeed seems to gradually allocate, but I'm not sure. 
@tarakby Could you maybe help us here? 🙏

Reply:
i'm just gonna do an integer division by 2, any objections?

Reply:
Some comments:

- I think `bytes = ceil(log256(maxVal))` doesn't cover all cases. Think of the case when `maxVal` is a power of 256. An easier example considering base `2` instead of base `256` : when `maxVal = 8`, the formula above gives 3 while the number of bits should be 4. 
A formula that works in all cases is `bytes = floor(log256(maxVal) + 1)`
- `math/big`  works in `word` instead of bytes. Words in our case are 64 bits (and that would be an overestimate in case we switch to a 32 bits cpu). So we would need to work in base 2^64 instead of 2^8. 
- as @Bastian mentioned, `math/big` over-allocates 4 words more than the size it needs. 
-  @dreamsmasher I agree, divisions by 2 or power of 2s are a good way to avoid float operations. We can use a similar trick and we can tighten the overestimate a bit more (to do better than 0.41 ~ 0.5)

I would suggest :
```
// maxVal = 10^l    , with l = |input|
// words = 4 + floor(log_(2^64)(maxVal) + 1)
//             = 4 +  floor(log_(2^64)(10^l) + 1) via substitution
//             = 4 +  floor(l * C) + 1   , with C = log_(2^64)(10)
//   over estimating C:  C = 0.05190512..  <= 213/4096
//. words = 5 + floor((l*213)/2^12))
//  bytes = words * 4

estimatedSize := 4 * (5+ ((len(input)*213) >> 12)
```

Please check the steps in case I missed something :)

Reply:
seems about right, I love the degree of bikeshedding in this thread.

Seems you can optimize the steps a bit more:
```
estimatedSize 
  = 4 * (5 + ((l * 213) >> 12))
  = 20 + (4 * ((l * 213) >> 12))
  = 20 + (((l * 213) >> 12) << 2)
  = 20 + ((l * 213) >> 10)
```

https://imgur.com/a/ItClhVp

Reply:
@tarakby Thank you!

Reply:
Guess this should go into OverEstimateBigIntFromString? https://github.com/onflow/cadence/blob/7a111462643d4e2f796704c2fcb8725249d33c1a/runtime/common/bigInt.go#L47

Reply:
@dreamsmasher We already have similar formulas in https://github.com/onflow/cadence/blob/master/runtime/common/metering.go, so that's a good spot to put it, along with a comment explaining the magic numbers

Reply:
I was so sure we already had this function, but couldn't find it, thank you for finding it.

So we already have this estimation function, how does the new one differ? Can we just use the existing one?

Reply:
it breaks a bunch of tests, I'll just use that function for now and then open a separate PR for the new bigInt estimation algo and tests.

Reply:
Agreed, use the existing function for this PR and then we can potentially open another PR to improve it. 
The discussion here is not bikeshedding, it's just solving the problem correctly, which involves figuring our how to meter memory usage, which is unfortunately not straight forward (well, we had already figured it out in the past, but forgot :D, then again it seems to have lead to an improvement, haha)

Reply:
hahaha, I'm calling it bikeshedding because everyone got nerdsniped by this problem. I asked on twitter a few days ago and got like 30 responses too.

Reply:
I didn't know we had the "old" formula already. The old and new formulas seem pretty much equivalent.  They are just 2 methods of the same reasoning (one uses base 2^64 and the other base 2). I don't think the new one is an improvement (unless you see big result divergence). It is fine to just keep the old formula and not update it.

Reply:
Great, thank you for the analysis @tarakby